### DISCUSSION TITLE
How to make reusable UI agnostic store/slice? (use case timer) I'm Migrating from Flutter/Dart

### DISCUSSION THREAD

**Discussion by MiniSuperDev** [pmndrs/zustand#2062] [Q&A]
Hello, how can I make a reusable timer with Zustand? (I want to share this state logic in react/vue/cli/etc)

I would also like to be able to have in my app several different instances of that timer, that is, several timers on screens at the same time, each one in with isolated state.

In dart/flutter I use bloc (similar to zustand/redux but for flutter)  but it use a class, so I create the ticker (like js `setInterval`) in this class (store/bloc), so also the ticker service is encapsulated in this class(store/bloc), and also it dispose / cancel / restart the subscription to the ticker when needed.

Like this:
https://github.com/felangel/bloc/blob/master/examples/flutter_timer/lib/timer/bloc/timer_bloc.dart

I like it this way because I create all the unit tests for this class, and now  I am just reusing it in any framework.

But with zustand I have this:

https://github.com/NacomiTagiera/Pomodoro-timer/blob/main/src/store/timerStore.ts

Where the timer is not created there, I only have the method  `countDown` and in the UI, in the Timer Component I use the `setInterval` in an `useEffect` to send this event.

Like this:

https://github.com/NacomiTagiera/Pomodoro-timer/blob/main/src/components/Timer.tsx#L19-L40

How can I achieve something similar to what I have in flutter/dart but in zustand?

I mean encapsulate all this logic in a reusable store/slice, to not add more logic in the UI or other place, because I want to test it in isolation, and simple reuse it in any UI framework without manage the setInterval outside the store.

And of course I want to be able to create multiple instances of the same slice/store/bloc.

This is just a simple case. Since in Dart/Flutter I have similar ones, so I want to know how to migrate them, but let the blocs/stores remain UI framework agnostic.

Thanks

**Comment by dai-shi** [comment]
While we could do something with middleware, an easy and recommended zustand way is to create a custom hook.

```js
// not for export
const useFooStoreBase = create((set) => ({
  // ...
}));

export const useFooStore = () => {
  const state = useFooStoreBase();
  useEffect(() => {
    // ...
  }, [/* ... */]);
  return /* ... */;
};
```
(omitting selectors, but they would probably be used in practice.)
